File: src/ocr/providers/openai.ts
import type { OCRModelRunRequest, OCRModelRunResult } from '../types';
import {
  buildPromptCacheKey,
  getOpenAIReasoningEffort,
  getRetryMaxOutputTokens,
  isLikelyTruncatedText,
} from '../provider-utils';
import { readTextFromOpenAIResponse } from '../text-readers';
import {
  isRecord,
  getRecordProperty,
  getStringProperty,
  getNumberProperty,
} from '../../lib/type-guards';

export async function runOpenAIOCR(params: {
  request: OCRModelRunRequest;
  apiKey: string;
  startedAt: number;
  maxOutputTokens: number;
}): Promise<OCRModelRunResult> {
  const { request, apiKey, startedAt, maxOutputTokens } = params;
  const promptCacheKey = buildPromptCacheKey(
    request.modelId,
    request.systemPrompt,
    request.userPrompt,
    request.pdfBase64
  );

  const runOpenAI = async (maxTokensForCall: number) => {
    const payload: Record<string, unknown> = {
      model: request.modelId,
      max_output_tokens: maxTokensForCall,
      input: [
        {
          role: 'system',
          content: [{ type: 'input_text', text: request.systemPrompt }],
        },
        {
          role: 'user',
          content: [
            {
              type: 'input_file',
              filename: request.filename || 'document.pdf',
              file_data: `data:application/pdf;base64,${request.pdfBase64}`,
            },
            {
              type: 'input_text',
              text: request.userPrompt,
            },
          ],
        },
      ],
      prompt_cache_key: promptCacheKey,
    };

    const reasoningEffort = getOpenAIReasoningEffort(request.modelId);
    if (reasoningEffort) {
      payload.reasoning = { effort: reasoningEffort };
    }

    const response = await fetch('https://api.openai.com/v1/responses', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    const data: unknown = await response.json();
    if (!isRecord(data)) {
      throw new Error('OpenAI request returned invalid response.');
    }
    if (!response.ok) {
      const errorObj = getRecordProperty(data, 'error');
      const errorMessage = errorObj ? getStringProperty(errorObj, 'message') : undefined;
      throw new Error(errorMessage ?? 'OpenAI request failed.');
    }
    return data;
  };

  let data = await runOpenAI(maxOutputTokens);
  let text = readTextFromOpenAIResponse(data);
  const status = getStringProperty(data, 'status') ?? '';
  const incompleteDetails = getRecordProperty(data, 'incomplete_details');
  const incompleteReason = incompleteDetails ? getStringProperty(incompleteDetails, 'reason') ?? '' : '';
  const truncatedByTokens = status === 'incomplete' && incompleteReason === 'max_output_tokens';

  if (truncatedByTokens || isLikelyTruncatedText(text)) {
    const retryMax = getRetryMaxOutputTokens(maxOutputTokens);
    if (retryMax > maxOutputTokens) {
      data = await runOpenAI(retryMax);
      text = readTextFromOpenAIResponse(data);
    }
  }

  const usageObject = getRecordProperty(data, 'usage');
  const promptTokensDetails = usageObject ? getRecordProperty(usageObject, 'prompt_tokens_details') : undefined;
  const cachedInputTokens = promptTokensDetails ? getNumberProperty(promptTokensDetails, 'cached_tokens') ?? 0 : 0;

  return {
    text,
    inputTokens: usageObject ? getNumberProperty(usageObject, 'input_tokens') ?? 0 : 0,
    outputTokens: usageObject ? getNumberProperty(usageObject, 'output_tokens') ?? 0 : 0,
    latencyMs: Date.now() - startedAt,
    cachedInputTokens,
    cacheHit: cachedInputTokens > 0,
    cacheWriteTokens: 0,
  };
}

File: src/ocr_mini_bench/ocr/providers/openai.py
"""OpenAI OCR provider. Mirrors `src/ocr/providers/openai.ts`.

Calls the `/v1/responses` endpoint with the PDF passed inline as a base64
`input_file`. Reasoning effort is selected per model id; truncation is retried
once with a larger `max_output_tokens` ceiling.
"""

from __future__ import annotations

import time
from typing import Any

import httpx

from ...lib.type_guards import (
    get_number_property,
    get_record_property,
    get_string_property,
    is_record,
)
from ..provider_utils import (
    build_prompt_cache_key,
    get_openai_reasoning_effort,
    get_retry_max_output_tokens,
    is_likely_truncated_text,
)
from ..text_readers import read_text_from_openai_response
from ..types import OCRModelRunRequest, OCRModelRunResult


async def run_openai_ocr(
    *,
    request: OCRModelRunRequest,
    api_key: str,
    started_at: float,
    max_output_tokens: int,
    client: httpx.AsyncClient,
) -> OCRModelRunResult:
    prompt_cache_key = build_prompt_cache_key(
        request.model_id, request.system_prompt, request.user_prompt, request.pdf_base64
    )
    filename = request.filename or "document.pdf"

    async def _call(max_tokens_for_call: int) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "model": request.model_id,
            "max_output_tokens": max_tokens_for_call,
            "input": [
                {
                    "role": "system",
                    "content": [{"type": "input_text", "text": request.system_prompt}],
                },
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "input_file",
                            "filename": filename,
                            "file_data": f"data:application/pdf;base64,{request.pdf_base64}",
                        },
                        {"type": "input_text", "text": request.user_prompt},
                    ],
                },
            ],
            "prompt_cache_key": prompt_cache_key,
        }

        reasoning_effort = get_openai_reasoning_effort(request.model_id)
        if reasoning_effort:
            payload["reasoning"] = {"effort": reasoning_effort}

        response = await client.post(
            "https://api.openai.com/v1/responses",
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            json=payload,
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise RuntimeError("OpenAI request returned invalid response.") from exc
        if not is_record(data):
            raise RuntimeError("OpenAI request returned invalid response.")
        if response.status_code >= 400:
            error_obj = get_record_property(data, "error")
            message = get_string_property(error_obj, "message") if error_obj else None
            raise RuntimeError(message or "OpenAI request failed.")
        return data

    data = await _call(max_output_tokens)
    text = read_text_from_openai_response(data)
    status = get_string_property(data, "status") or ""
    incomplete_details = get_record_property(data, "incomplete_details")
    incomplete_reason = (
        get_string_property(incomplete_details, "reason") or "" if incomplete_details else ""
    )
    truncated_by_tokens = status == "incomplete" and incomplete_reason == "max_output_tokens"

    if truncated_by_tokens or is_likely_truncated_text(text):
        retry_max = get_retry_max_output_tokens(max_output_tokens)
        if retry_max > max_output_tokens:
            data = await _call(retry_max)
            text = read_text_from_openai_response(data)

    usage = get_record_property(data, "usage")
    prompt_tokens_details = (
        get_record_property(usage, "prompt_tokens_details") if usage else None
    )
    cached_input_tokens = (
        int(get_number_property(prompt_tokens_details, "cached_tokens") or 0)
        if prompt_tokens_details
        else 0
    )
    input_tokens = int(get_number_property(usage, "input_tokens") or 0) if usage else 0
    output_tokens = int(get_number_property(usage, "output_tokens") or 0) if usage else 0

    return OCRModelRunResult(
        text=text,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        latency_ms=int((time.time() - started_at) * 1000),
        cached_input_tokens=cached_input_tokens,
        cache_hit=cached_input_tokens > 0,
        cache_write_tokens=0,
    )
